packages/ts-client/src/webrtc/transciever.ts

```diff
@@ -0,0 +1,210 @@
+        encoding.active = true;
+      } else {
+        disabledTrackEncodings.push(encoding.rid! as TrackEncoding);
+      }
+    });
+    disabledTrackEncodingsMap.set(trackContext.trackId, disabledTrackEncodings);
+  } else {
+    transceiverConfig = {
+      direction: 'sendonly',
+      sendEncodings: [
+        {
+          active: true,
+        },
+      ],
+      streams: trackContext.stream ? [trackContext.stream] : [],
+    };
+  }
+
+  if (trackContext.maxBandwidth && transceiverConfig.sendEncodings)
+    applyBandwidthLimitation(
+      transceiverConfig.sendEncodings,
+      trackContext.maxBandwidth,
+    );
+
+  return transceiverConfig;
+};
+
+export const setTransceiversToReadOnly = (connection: RTCPeerConnection) => {
+  connection
+    .getTransceivers()
+    .forEach((transceiver) => (transceiver.direction = 'sendonly'));
+};
+
+type Mid = string;
+type TrackId = string;
+type MidToTrackId = Record<Mid, TrackId>;
+
+export const getMidToTrackId = <EndpointMetadata, TrackMetadata>(
+  connection: RTCPeerConnection | undefined,
+  localTrackIdToTrack: Map<
+    RemoteTrackId,
+    TrackContextImpl<EndpointMetadata, TrackMetadata>
+  >,
+  midToTrackId: Map<string, string>,
+  localEndpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata>,
+): MidToTrackId | null => {
+  if (!connection) return null;
+
+  // - negotiated unmuted tracks: tracks added in previous negotiation, data is being transmitted
+  // - not yet negotiated tracks: tracks added in this negotiation, data will be transmitted after successful negotiation
+  const mappingFromTransceivers = getTransceiverMapping(
+    connection,
+    localTrackIdToTrack,
+  );
+
+  // - negotiated unmuted tracks: tracks added in previous negotiation, data is being transmitted
+  // - negotiated muted tracks: tracks added in previous negotiation, data is not being transmitted but can be transmitted in the future
+  const mappingFromLocalNegotiatedTracks = getAllNegotiatedLocalTracksMapping(
+    midToTrackId,
+    localEndpoint,
+  );
+
+  return { ...mappingFromTransceivers, ...mappingFromLocalNegotiatedTracks };
+};
+
+const getTrackContext = <EndpointMetadata, TrackMetadata>(
+  localTrackIdToTrack: Map<
+    RemoteTrackId,
+    TrackContextImpl<EndpointMetadata, TrackMetadata>
+  >,
+  localTrackId: string,
+) =>
+  Array.from(localTrackIdToTrack.values()).find(
+    (trackContext) => trackContext?.track?.id === localTrackId,
+  )!;
+
+const getTransceiverMapping = <EndpointMetadata, TrackMetadata>(
+  connection: RTCPeerConnection,
+  localTrackIdToTrack: Map<
+    RemoteTrackId,
+    TrackContextImpl<EndpointMetadata, TrackMetadata>
+  >,
+): MidToTrackId =>
+  connection
+    .getTransceivers()
+    .filter((transceiver) => transceiver.sender.track?.id && transceiver.mid)
+    .reduce(
+      (acc, transceiver) => {
+        const localTrackId = transceiver.sender.track!.id;
+        const mid = transceiver!.mid!;
```

Comment:
It would require a little more work, but in future refactor it would be nice to create filter functions that returns proper TS type, so no `!` will be required.